File: fastapi-server/src/modules/user/application/commands/create_user.py
"""user モジュールのユーザー作成コマンド。"""

from src.modules.user.domain import Email, UserDomainService, UserName
from src.shared.events import IEventDispatcher

from ..dto import CreateUserRequest, CreateUserResponse


class CreateUserCommand:
    """ユーザー作成コマンド

    ドメインサービスを使用してユーザーを作成します。
    メールアドレスの一意性チェックが自動的に行われます。

    例外:
        UserValidationError: バリデーションに失敗した場合
        DuplicateEmailError: メールアドレスが既に存在する場合
    """

    def __init__(
        self,
        domain_service: UserDomainService,
        event_dispatcher: IEventDispatcher | None = None,
    ) -> None:
        """コマンドを初期化

        Args:
            domain_service: ユーザードメインサービス
            event_dispatcher: イベントディスパッチャー（オプション）
        """
        self._domain_service = domain_service
        self._event_dispatcher = event_dispatcher

    async def execute(self, request: CreateUserRequest) -> CreateUserResponse:
        """ユーザー作成コマンドを実行

        Args:
            request: 作成リクエスト DTO

        Returns:
            作成レスポンス DTO

        Raises:
            UserValidationError: バリデーションに失敗した場合
            DuplicateEmailError: メールアドレスが既に存在する場合
        """
        # 値オブジェクトを作成（バリデーションが行われる）
        email = Email(request.email)
        name = UserName(
            first_name=request.first_name,
            last_name=request.last_name,
        )

        # ドメインサービスを使用してユーザーを作成
        user = await self._domain_service.create_user(
            email=email,
            name=name,
            bio=request.bio,
        )

        # ドメインイベントをディスパッチ
        if self._event_dispatcher:
            events = user.get_domain_events()
            await self._event_dispatcher.dispatch_all(events)
            user.clear_domain_events()

        return CreateUserResponse(
            id=user.id,
            email=str(user.email),
            full_name=str(user.name),
            created_at=user.created_at,
        )

File: go-server/internal/modules/user/application/commands/create_user.go
package commands

import (
	"context"

	"cariban/go-server/internal/modules/user/application"
	"cariban/go-server/internal/modules/user/domain"
	"cariban/go-server/internal/shared/events"
)

type CreateUserCommand struct {
	service    *domain.UserDomainService
	dispatcher events.Dispatcher
}

func NewCreateUserCommand(service *domain.UserDomainService, dispatcher events.Dispatcher) *CreateUserCommand {
	return &CreateUserCommand{service: service, dispatcher: dispatcher}
}

func (c *CreateUserCommand) Execute(ctx context.Context, req application.CreateUserRequest) (application.CreateUserResponse, error) {
	email, err := domain.NewEmail(req.Email)
	if err != nil {
		return application.CreateUserResponse{}, err
	}
	name, err := domain.NewUserName(req.FirstName, req.LastName)
	if err != nil {
		return application.CreateUserResponse{}, err
	}

	user, err := c.service.CreateUser(ctx, email, name, req.Bio)
	if err != nil {
		return application.CreateUserResponse{}, err
	}

	if c.dispatcher != nil {
		if err := c.dispatcher.DispatchAll(ctx, user.DomainEvents()); err != nil {
			return application.CreateUserResponse{}, err
		}
		user.ClearDomainEvents()
	}

	return application.CreateUserResponse{
		ID:        user.ID,
		Email:     user.Email.String(),
		FullName:  user.Name.FullName(),
		CreatedAt: user.CreatedAt,
	}, nil
}
